runtime/lua/vim/lsp/util.lua

```diff
@@ -121,50 +121,61 @@ local function get_line_byte_from_position(bufnr, position)
 end
 
 --- Process and return progress reports from lsp server
-function M.get_progress_messages()
+function M.get_progress_messages(bufnr)
 
   local new_messages = {}
   local msg_remove = {}
   local progress_remove = {}
 
+  local uri = vim.uri_from_bufnr(bufnr)
+
+  -- TODO: add function argument to specify a client id
   for _, client in ipairs(vim.lsp.get_active_clients()) do
-      local messages = client.messages
-      local data = messages
-      for token, ctx in pairs(data.progress) do
-
-        local new_report = {
-          name = data.name,
-          title = ctx.title or "empty title",
-          message = ctx.message,
-          percentage = ctx.percentage,
-          progress = true,
-        }
-        table.insert(new_messages, new_report)
+    local messages = client.messages
+    local data = messages
+
+    for token, ctx in pairs(data.progress) do
+      local new_report = {
+        name = client.name,
+        id = client.id,
+        title = ctx.title or "empty title",
+        message = ctx.message,
+        percentage = ctx.percentage,
+        progress = true,
+        done = ctx.done,
```

Comment:
Where is "done" is the LSP spec? https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workDoneProgressReport